### PR TITLE
feature/Add-celery-and-rabbitmq

### PR DESCRIPTION
**Pull request by jsarrolt**

### CODE REVIEW

**Comment by jsarrolt** [comment]
While running rabbit, i get this warning

image

With `rabbitmq.conf` option must be fixed, but is not working. I keep it since it seems to be fix soon. What do you think @DanTcheche ?

https://github.com/rabbitmq/rabbitmq-server/discussions/11978

File: app/api/v1/endpoints/users.py [hunk 1]
```diff
@@ -1,13 +1,26 @@
...
 router = APIRouter()
 
 
 @router.get("/current", status_code=status.HTTP_200_OK)
-def get_current_patient(
+def get_current_user(
     current_user: CurrentUser,
 ) -> UserResponse:
-    return UserResponse.model_validate(current_user)
+    user = UserResponse.model_validate(current_user)
```
Comment (DanTcheche) [inline app/api/v1/endpoints/users.py:17]:
```suggestion
    return UserResponse.model_validate(current_user)
```

File: app/celery/tasks/emails.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+from app.services.users_service import UsersService
+
+from app.repositories.users_repository import users_repository
+
+
+@celery.task
+def send_reminder_email() -> None:
+    session = SessionLocal()
+    try:
+        users = UsersService(session, users_repository).list(ListFilter())
```
Comment (DanTcheche) [inline app/celery/tasks/emails.py:15]: Why empty list filter?

File: app/celery/tasks/emails.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        for user in users.data:
+            EmailService(ExampleEmailClient()).send_user_remind_email(
+                user.__dict__
```
Comment (DanTcheche) [inline app/celery/tasks/emails.py:18]:
```suggestion
                user.model_dump()
```

File: app/clients/example_email_client.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        super().__init__(client=None)
+
+    def send_email(
+        self,
+        to_emails: List[str],
+        html_message: str,
+        set_configuration_name: bool = False,
+    ) -> None:
+        for email in to_emails:
+            logger.info(f"Sending email to {email}")
```
Comment (DanTcheche) [inline app/clients/example_email_client.py:21]: Not HIPAA compliant to log emails.

File: app/services/emails_service.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        return self.email_template.substitute(
+            subject=subject,
+            html_message=html_message,
+            sender_email=settings.SENDER_EMAIL,
+            recipient_email=recipient_email,
+        )
+
+    def send_new_user_email(
+        self,
+        user: dict,
```
Comment (DanTcheche) [inline app/services/emails_service.py:49]: Please use schemas not dicts.

**Review by DanTcheche — CHANGES_REQUESTED** [review]
Nice job!

File: app/celery/tasks/emails.py (hunk 2 above)
Comment (jsarrolt) [reply app/celery/tasks/emails.py]: Is mandatory to pass it and here is just an example. Maybe i can add explicit page and page_size so it is clearer

File: app/celery/tasks/emails.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            )
+    finally:
+        session.close()
+
+
+@celery.task
+def send_welcome_email(created_user: dict) -> None:
+    session = SessionLocal()
+    try:
+        EmailService(ExampleEmailClient()).send_new_user_email(
```
Comment (DanTcheche) [inline app/celery/tasks/emails.py:31]: We can go a step further here.
What if sending the email or this task fails?
How can we make sure that this task is retried until the email is sent?

Investigate a little bit and if nothing comes to mind contact me 😄 .

**Review by DanTcheche — CHANGES_REQUESTED** [review]

File: app/celery/tasks/emails.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        for user in users.data:
+            EmailService(ExampleEmailClient()).send_user_remind_email(
+                UserInDB.model_validate(user)
+            )
+    finally:
+        session.close()
+
+
+@celery.task(
+    autoretry_for=(ExternalProviderException,),
```
Comment (DanTcheche) [inline app/celery/tasks/emails.py:34]:
```suggestion
    autoretry_for=(EmailClientException,),
```

File: app/celery/tasks/emails.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    retry_backoff=True,
```
Comment (DanTcheche) [inline app/celery/tasks/emails.py:35]: Where can we define the amount of time between retries?
I would set that up in a env variable too, probably 30 segs.

File: app/use_cases/users/create_user_use_case.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    def __init__(self, session: Session):
+        self.session = session
+
+    def execute(self, create_user_request: CreateUserRequest) -> UserResponse:
+        from app.celery.tasks.emails import send_welcome_email
+
+        users_service = UsersService(self.session, users_repository)
+        if users_service.get_by_email(create_user_request.email):
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
```
Comment (DanTcheche) [inline app/use_cases/users/create_user_use_case.py:20]: I would use 409 here, it's not the users request fault.

**Review by DanTcheche — CHANGES_REQUESTED** [review]

**Comment by jsarrolt** [comment]
image

**Review by DanTcheche — APPROVED** [review]

**Pull request merged**